Switch from pg_local to pg

mkuratczyk opened the pull request:
pg_local is based on the pg2 module, which got removed
from Erlang/OTP years ago. It was replaced by the more
efficient pg module, so let's use it directly.

We use node-local scopes and single-item groups so that
(de)registration is fast, but we can list all connections/channels
(by listing all groups in the scope).

Initially we wanted to remove pg_local altogether, but we decided to keep it
just for rabbit_direct, which uses potentially remote PIDs, so it can't use pg.

mkuratczyk committed:
Remove rabbit_stream:kill_connection/1

This function was not called from anywhere
and was the only place that expected stream
connections to be registered in pg_local.

acogoluegnes committed:
Retry connection attempt in stream test

The metadata can take a few milliseconds to reflect a node is down, so
the test code must consider a connection failure and test another node.

mkuratczyk committed:
Delete pg_local module

mkuratczyk committed:
Switch from pg_local to pg

pg_local is based on the pg2 module, which got removed from Erlang/OTP years ago. It was replaced by the more efficient pg module, so use it directly for AMQP 0-9-1 channels, TCP and non-AMQP connection tracking, and related tests.

We use node-local scopes and single-item groups so that (de)registration is fast, but we can list all connections and channels by listing all groups in the scope.

rabbit_direct still uses pg_local for one more commit; the next commit moves it to a local ETS table because pg cannot represent those peers in all deployment topologies (see that commit message).

mkuratczyk committed:
rabbit_direct: track connections in ETS instead of pg

pg only allows registering processes that are node local.
Direct connections pass the client pid, which can be remote.
Therefore, here we switch from pg_local to ETS, instead of pg.

mkuratczyk committed:
Add CT for rabbit_direct list and list_local

ansd commented on deps/rabbit/src/rabbit_volatile_queue.erl:
```diff
@@ -208,7 +208,7 @@ local_call(Pid, Request) ->
 is_local(Pid) ->
     rabbit_amqp_session:is_local(Pid) orelse
-    pg_local:in_group(rabbit_channels, Pid).
+    pg:get_local_members(rabbit:pg_scope_amqp091_channel(), Pid) =/= [].
```
Should we follow the same approach as done for `rabbit_amqp_session:is_local(Pid)` and introduce a function `rabbit_channel:is_local/1`?
Currently, this code line leaks a lot of details of how channels are tracked into the volatile queue type.

ansd commented on deps/rabbit/src/rabbit_direct.erl:
```diff
@@ -186,7 +190,7 @@ connect1(User = #user{username = Username}, VHost, Pid, Infos) ->
             try rabbit_access_control:check_vhost_access(User, VHost,
                                                {ip, PeerHost}, AuthzContext) of
-                ok -> ok = pg_local:join(rabbit_direct, Pid),
+                ok -> ets:insert(?TABLE, {Pid}),
```
It's unfortunate that we can't use pg here. Does this Pid need to be monitored? In other words, is disconnect/2 always guaranteed to be called? Just want to make sure we won't create a leak here.

michaelklishin replied:
`is_disconnect/2` is seemingly only run when `amqp_direct_connection:terminate/2` runs. Which is not the case in certain failure scenarios, e.g. on the node which is the client fails.

We can introduce an equivalent of the metrics GC process which runs periodically and sweeps what seems stale (unused).

ansd replied:
Alternatively, we don't delete the pg_local module and just keep using it only in `rabbit_direct`. This sounds better to me than manual ETS tables with periodically sweeping them.

This whole PR here still provides benefits by removing pg_local usages almost everywhere. We can mark pg_local as deprecated via code so that no new code starts using it.

michaelklishin replied:
I agree, we can keep that module for just `rabbit_direct` and use the standard `pg` as much as we can.

mkuratczyk committed:
Bring pg_local back for rabbit_direct

Revert changes to rabbit_direct. Don't remove
pg_local completely - leave it just to be used by rabbit_direct.

mkuratczyk committed:
Add rabbit_channel:is_local/1

mkuratczyk commented:
PR review feedback has been incorporated: I've re-introduced pg_local just for rabbit_direct and added `rabbit_channel:is_local/1`

ansd committed:
Mark pg_local as deprecated

This way new code won't accidentally use this module.

ansd commented:
I added a deprecation directive to deprecate the entire `pg_local` module. This way, new code won't accidentally use this old module.
